Soft shutdown review comments

BadSingleton opened the pull request.

BadSingleton committed:
Remove unused code

Addresses comment:
amos402: Hummm, seems I only use it in my previous version, I will
check it latter.
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r343815280

BadSingleton committed:
Fixes some typos in TestDomainReload

Addresses comments:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390639971
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390640042
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390643563

BadSingleton committed:
Adds missing assert

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390644837

BadSingleton committed:
Adds code comments to CrossDomainObject test

Addresses comment:
lostmu: Add a comment explaining the intent of this method or extract
this code into a separate method with clear name.
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390644281

BadSingleton committed:
Adds documentation to TestDomainReload.TestClassReference

Addresses comment:
lostmu: BTW, it would be helpful to add comments with equivalent Python
code like [...]
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390650653

BadSingleton committed:
Inline called-once methods in TestDomainReload

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390653941

BadSingleton committed:
Adds numbering to domain names in TestDomainReload

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390655505

BadSingleton committed:
Use Console.WriteLine formatting overload

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390656145

BadSingleton committed:
Simplify PythonRunner.RunPython

Addresses comments:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390656459
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390657130

BadSingleton committed:
Release the GIL

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390659969

BadSingleton committed:
BorrowReference instead of increasing the refcount

Addresses comments:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390660744
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390660804

BadSingleton committed:
Add IsTypeObject to Managed type

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390663522

BadSingleton committed:
Rename PyObjectSerializeContext

to InterDomainContext

Addresses Comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390664306

BadSingleton committed:
Rename StashPush/Pop methods

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390669351

BadSingleton committed:
Rename ClassManager.RemoveClasses

To `DisposePythonWrappersForClrTypes`

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390670768

BadSingleton committed:
Rename ClassManager.OnVisit

To `TraverseTypeClear` to emphasis it's used for type traversal
(similar to tp_traverse)

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390671251

BadSingleton committed:
fixup! Inline called-once methods in TestDomainReload

BadSingleton committed:
Change refcount logic in ImportHook.Shutdown

If the shutdown mode is Reload, don't decref.

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390685930

BadSingleton committed:
Remove TODOS

tp_clear and tp_traverse always return 0. No error check to do

Addresses comments:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390688176
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390688190

BadSingleton committed:
Incref a valid pointer

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390689745

BadSingleton committed:
Use Py_CLEAR

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390690092

BadSingleton committed:
Fixes GIL grabbing during init and shutdown

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390694298

BadSingleton committed:
Add a parameter to specify the shutdown mode.

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390659648

BadSingleton committed:
Add typeoffset.cs

Move the partial class definition of TypeOffset from interop.cs to its
own file.

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390686331

BadSingleton committed:
Extract utility method in TestDomainReload

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390645914

BadSingleton committed:
Add cleanup to ClassInfo

Addresses comment:
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r390671982

lostmsu commented on src/embed_tests/pyinitialize.cs:
```diff
@@ -24,10 +24,12 @@ public static void StartAndStopTwice()
-                Runtime.Runtime.XIncref(argv.Handle);
-                Assert.AreNotEqual(0, argv.Length());
+                var argvref = new BorrowedReference(Runtime.Runtime.PySys_GetObject("argv"));
+                using(var argv = new PyList(argvref.DangerousGetAddress()))
```
Rather than adding `DangerousGetAddress` here, we should have an `internal PyList(BorrowedReference)` constructor.

lostmsu commented on src/embed_tests/pyinitialize.cs:
```diff
@@ -24,10 +24,12 @@ public static void StartAndStopTwice()
             {
-                Runtime.Runtime.XIncref(argv.Handle);
-                Assert.AreNotEqual(0, argv.Length());
+                var argvref = new BorrowedReference(Runtime.Runtime.PySys_GetObject("argv"));
```
NIT: I believe `PySys_GetObject` is rarely used. It's return type can be changed to `BorrowedReference`.

BadSingleton replied:
It's used around 15 times in the codebase and most places use the IntPtr. Best be done a separate PR

amos402 replied:
You removed XIncref which means your reference count will be incorrect.
Why make these changes, there's no bugs here.

BadSingleton replied:
Ah my bad. Thinking about it, why do we need to increment the refcount of a borrowed reference? I added a new constructor to make `PyList` from a `BorrowedReference`. The list does take ownership of it (and decrefs it when Disposed)

lostmsu commented on src/runtime/classmanager.cs:
```diff
@@ -76,32 +71,32 @@ private static int OnVisit(IntPtr ob, IntPtr arg)
             foreach (var cls in cache.Values)
             {
                 // This incref is for cache to hold the cls,
-                // thus no need for decreasing it at StashPop.
+                // thus no need for decreasing it at RestoreRuntimeData.
```
Does this mean whatever is stored in `storage` leaks every domain restart?

BadSingleton replied:
If I understand the code correctly, at shutdown, the types are cleared. To avoid them actually going away, we incref them. Why doesn't this leak? Because this incref is equivalent to the new reference received when creating the type (except is is never created when reloading a domain, only preserved)
@amos402  is this correct?

amos402 replied:
Correct. The types' data will be recovering at the next startup.
Yet there's still not having any error fallback now. The reload mode is not a completed mode right now.

lostmsu commented on src/runtime/importhook.cs:
```diff
@@ -107,29 +107,30 @@ internal static void Shutdown()
-            Runtime.XDecref(py_clr_module);
-            py_clr_module = IntPtr.Zero;
-            if (shouldFreeDef)
+#if !NETSTANDARD
+            if(Runtime.ShutdownMode != ShutdownMode.Reload)
+#endif
```
3.0 is supposed to only be .NET Standard, so I think this is incorrect. I suspect this should instead be a runtime check if `AppDomain` is fully supported.

BadSingleton replied:
I don't think AppDomain needs to be supported. I've not done it but if the domain is started/stopped by the mono runtime at C/C++ level that counts too, no?

amos402 replied:
Why add a condition branch, you can just keep its sole purpose from disturbing the shutdown mode. Also for the same when you save their state, just segregate them.

lostmsu commented on src/runtime/runtime.cs:
```diff
@@ -148,7 +149,8 @@ internal static void Initialize(bool initSigs = false, ShutdownMode mode = Shutd
             else
             {
-                PyGILState_Ensure();
+                PyEval_InitThreads();
```
This needs some explanation. Also, according to [this SO](https://stackoverflow.com/questions/15470367/pyeval-initthreads-in-python-3-how-when-to-call-it-the-saga-continues-ad-naus), `PyEval_InitThreads` acquires GIL, so I am not sure what to think about `PyGILState_Ensure` in the next line.

amos402 replied:
This change will occur a bug by removing the `PyGILState_Ensure`.
GIL released on the previous shutdown, it needs GIL here.
`PyEval_InitThreads` is not the necessary one, it initialized at previous startup.

BadSingleton replied:
Looking at Python's code (ceval.c), `PyEval_InitThreads` only acquires the GIL if the GIL hasn't been created yet (also creates it). Calling it more than once is a no-op, and does not acquire it. `PyGILState_Ensure` is required so that the GIL is effectively acquired. @amos402 is right that the `PyEval_InitThreads` call is not required. I'll remove it and add a code comment.

lostmsu commented:
BTW, @BadSingleton this seems to be a good way to get fixes for the PR. I only wish you tagged me earlier, so I could review.

@amos402 , do you have time to review this? If not, I can create a mirror branch in the upstream repo. It is a bit inconvenient, as the new branch would have to be reviewed from scratch, but I'd like to have this move quickly.

amos402 commented:
These changes didn't provide too much essence changes, it will be easier to track these trivial changes here.
Also, I'm still waiting for this pr be passing the CI.

BadSingleton commented:
@lostmsu The idea was to get this branch merged into `soft-shutdown` so the fixes could be reviewed with the existing PR in the main repo. Also this PR is best reviewed on a per-commit basis as each commit addresses one review comment (or two if they were similar).

For now the changes are small because I've addressed the review comments that seemed to be easy. I'm almost done with those, so expect heavier changes in the coming days/weeks.

BadSingleton committed:
Merge remote-tracking branch 'amos/soft-shutdown' into soft-shutdown-review-comments

BadSingleton committed:
fixup! Add cleanup to ClassInfo

amos402 commented on src/runtime/classmanager.cs:
```diff
@@ -481,5 +474,28 @@ internal ClassInfo()
+
+        ~ClassInfo()
+        {
+            Dispose();
```
It's not safe to call such a GIL needed method in GC thread, besides, its reference count already balanced, thus the `Dispose` is not necessary.
https://github.com/amos402/pythonnet/blob/2b7bcacb63b25fb760e38bc558e40cedd3351933/src/runtime/classmanager.cs#L224

BadSingleton replied:
Then the fix is to document that ClassInfo should not be used outside of ClassManager and that the caller is responsible to decrease the reference count of the members.

amos402 replied:
It didn't fix any things, it made bugs. This change will occur crash eventually
Check here:
https://github.com/amos402/pythonnet/blob/2b7bcacb63b25fb760e38bc558e40cedd3351933/src/runtime/classmanager.cs#L217-L225

amos402 replied:
ClassInfo is merely a temporary object for carrying the data. Of course, you can clear the members after using it although it won't influence the usages.
And the type of `member` can change into a generic type, it's some old code so it didn't use generic type. Well, but seems it's not necessary to change it here.

BadSingleton replied:
I will revert this change and add documentation instead

amos402 commented on src/runtime/pythonengine.cs:
```diff
@@ -330,11 +328,26 @@ public static void Shutdown()
             PyScopeManager.Global.Clear();
             ExecuteShutdownHandlers();
+            // Remember to shut down the runtime.
+            Runtime.Shutdown(mode);
```
`Runtime.Shutdown` called in `ExecuteShutdownHandlers`

BadSingleton replied:
I removed it as a ShutdownHandler and put it here, so that if the runtime was Initialized on Reload, but we want to shutdown with a Normal mode because the application is quitting, then we can do so.

For example, in the Unity Editor the `ShutdownMode` would always be `Reload`. Eventually, when the application quits, we want to completely shutdown the runtime.

amos402 commented on src/runtime/runtime.cs:
```diff
@@ -360,7 +365,7 @@ internal static void Shutdown()
                     // Some clr runtime didn't implement GC.WaitForFullGCComplete yet.
                 }
-                PyEval_SaveThread();
+                PyGILState_Release(state);
```
Just keep using `PyEval_SaveThread`, you need to release the GIL definitely here.

BadSingleton replied:
I thought we needed to keep the return value of `PyEval_SaveThread`. Took a look at cpython's source code, we do not.

amos402 commented:
Excepted for the type errors, did other rename behaviors have any contexts about why?

BadSingleton commented:
The context for every change is in the commit messages. It has a link to the review comment it addresses.

BadSingleton committed:
Add a new PyList constructor

To build from a BorrowedReference. Assumes Ownership.

BadSingleton committed:
Revert previous commit "Add cleaunp to class info"

And change the scope of the ClassInfo class and add some documentation.

BadSingleton committed:
Add corrections to the GIL acquiring/releasing

In Runtime's Initialize and Shutdown.

BadSingleton committed:
Factor out ExtensionType's Gc setup mechanic

To avoid code duplication.

amos402 commented on src/embed_tests/TestRuntime.cs:
```diff
@@ -40,16 +40,27 @@ public static void PlatformCache()
+            {
+                if(state != IntPtr.Zero)
+                {
+                    Runtime.Runtime.PyGILState_Release(state);
```
https://github.com/pythonnet/pythonnet/pull/958/#discussion_r478812048

> 
> 
> `Py_Finalize` will release the GIL, if you call `PyGILState_Release` after finalized, a fatal error would occur.

amos402 commented on src/runtime/methodbinding.cs:
```diff
@@ -22,11 +22,11 @@ public MethodBinding(MethodObject m, IntPtr target, IntPtr targetType)
             {
                 targetType = Runtime.PyObject_Type(target);
             }
+            Runtime.XIncref(targetType);
```
`PyObject_Type` returns new reference.

amos402 commented on src/runtime/runtime.cs:
```diff
@@ -193,6 +198,10 @@ internal static void Initialize(bool initSigs = false, ShutdownMode mode = Shutd
             AssemblyManager.UpdatePath();
+            if (state != IntPtr.Zero)
+            {
+                PyGILState_Release(state);
```
Why release the GIL, it supposed ready to run Python functions after initializing, with GIL. Compare to a normal runtime after calling Py_Initialize.

amos402 commented on src/runtime/runtime.cs:
```diff
@@ -306,17 +315,16 @@ private static IntPtr Get_PyObject_NextNotImplemented()
         }
 
-        internal static void Shutdown()
+        internal static void Shutdown(ShutdownMode mode)
```
If you tend to support asymmetric mode different from the initialization, you should add more validations. e.g. Normal mode can't upgrade to soft mode, it didn't save the recovery information in the first place.

BadSingleton replied:
Good point, I'll do that

amos402 commented on src/runtime/importhook.cs:
```diff
@@ -107,29 +107,30 @@ internal static void Shutdown()
+        internal static void SaveRuntimeData(RuntimeDataStorage storage)
         {
-            Runtime.XIncref(py_clr_module);
-            Runtime.XIncref(root.pyHandle);
```
https://github.com/amos402/pythonnet/pull/2/files/d2408b9f75484ebba21e40acca0c4da909697bc7#r478813946

amos402 commented on src/runtime/typeoffset.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Python.Runtime
+{
+    internal static partial class TypeOffset
```
Other offset types still in the `interop.cs`, why just extracts it alone. I think it's not a necessary change for shutdown mode features, it only increasing the changes for merging. Maybe you can bring these on upstream at first.

BadSingleton replied:
Good point, I'll revert that

BadSingleton committed:
Revert "Release the GIL"

This reverts commit 7ec9a6c677cd346afd3b75fe8022f51e4e076809.

BadSingleton committed:
fixup! Incref a valid pointer

BadSingleton committed:
Revert "Add typeoffset.cs"

This reverts commit f575bd329913c0791aa52b906ef5b6fc0865251e.

BadSingleton committed:
A better fix for releasing a re-acquiring the GIL

On shutdown and startup.
The main python thread always havea state. After releasing the state on
shutdown, restore the state of the main thread (thus also locking the
GIL).

BadSingleton committed:
Add validation for downgrading the shutdown mode

BadSingleton committed:
Merge remote-tracking branch 'amos/soft-shutdown' into soft-shutdown-review-comments

BadSingleton committed:
Fixes for the merge

BadSingleton committed:
Revert "Change refcount logic in ImportHook.Shutdown"

This reverts commit 82034dcb503bae528eb1c112017015f9d0410532.

BadSingleton committed:
Document refcount increase

BadSingleton committed:
Disambiguate Initialization

between initializing from CPython and initializing after a domain reload
Add an optional argument to signal to the initialization that it's
coming from CPython
